File: Automation/src/test/java/homework6/BaseTestClass4U.java
package homework6;

import BaseTest.BaseTest;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BaseTestClass4U extends BaseTest {
    static final String BASEURL = "https://4u.am/en/";
    private WebElement header;

    protected void scrollTo(WebElement element){
        JavascriptExecutor js = (JavascriptExecutor) driver;
//        js.executeScript("arguments[0].scrollIntoView();", element);
        js.executeScript("window.scrollBy("+element.getLocation().getX()+","+(element.getLocation().getY()+100)+")", "");
    }

    protected int getShopcartItemsCount(){
//        return driver.findElement(By.xpath("//div[@id=\"productQuantity\"]")).getText().equals("")?0: Integer.parseInt(driver.findElement(By.xpath("//div[@id=\"productQuantity\"]")).getText());
        driver.findElement(By.xpath("//div[@id=\"basketIcon\"]")).click();
        WebElement closeButton = new WebDriverWait(driver, Duration.ofSeconds(15)).until(ExpectedConditions.elementToBeClickable(By.xpath("//button[@id=\"basketCloseButton\"]")));
//        int count = new WebDriverWait(driver, Duration.ofSeconds(15)).until(ExpectedConditions.presenceOfAllElementsLocatedBy(By.xpath("//div[@class='basket_content_item']"))).size();
        int count = driver.findElements(By.xpath("//div[@class='basket_content_item']")).size();
//        driver.findElement(By.xpath("//button[@id=\"basketCloseButton\"]")).click();
        closeButton.click();
        return count;
    }

    protected void click(WebElement elem, WebElement header) {
        new Actions(driver).scrollByAmount(0,elem.getRect().y-header.getRect().height).perform();
        elem.click();
    }

    protected void click(WebElement elem) {
        if (header==null) {
            header = driver.findElement(By.xpath("//header"));
        }
        click(elem,header);
    }

    protected boolean checkItemExistsInShopcart(WebElement elem){
        driver.findElement(By.xpath("//div[@id=\"basketIcon\"]")).click();

        WebElement closeButton = new WebDriverWait(driver, Duration.ofSeconds(15)).until(ExpectedConditions.elementToBeClickable(By.xpath("//button[@id=\"basketCloseButton\"]")));
        List<WebElement> shopcartItems = driver.findElements(By.xpath("//div[@class=\"basket_content_item\"]"));
//        closeButton.click();

        for (WebElement item: shopcartItems) {
            if (itemToValue(elem).equals(basketItemToValue(item))) {
                closeButton.click();
                return true;
            }
        }

        closeButton.click();
        return false;
    }

    private Map<String,String> itemToValue(WebElement elem) {
        Map<String, String> itemProperties = new  HashMap<>();
        itemProperties.put("name",elem.findElement(By.xpath(".//p/a")).getText());
        itemProperties.put("price",elem.findElement(By.xpath(".//div/p/span/span")).getText());

        return itemProperties;
    }

    private Map<String,String> basketItemToValue(WebElement elem) {
        Map<String, String> itemProperties = new  HashMap<>();
        itemProperties.put("name",elem.findElement(By.xpath("//div[@class=\"basket_content_item\"][1]//h4")).getText());
        itemProperties.put("price",elem.findElement(By.xpath("//div[@class=\"basket_content_item\"][1]//p/span[1]")).getText());

        return itemProperties;
    }
}

File: homework6/BaseTestClass4U.py
from typing import List, Dict, overload

import pytest
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as ec
from selenium.webdriver.common.actions import action_builder
from selenium.webdriver import ActionChains


@pytest.mark.usefixtures("chrome_driver")
class BaseTestClass4U:
    driver: webdriver.Chrome

    BASE_URL = "https://4u.am/en/"
    header: WebElement
    basket_icon: By = (By.XPATH, "//div[@id='basketIcon']")

    @property
    def shopcart_items_count(self) -> int:
        self.driver.find_element(*self.basket_icon).click()
        close_button: WebElement = WebDriverWait(self.driver, 15).until(ec.element_to_be_clickable((By.XPATH, "//button[@id=\"basketCloseButton\"]")))
        count: int = len(self.driver.find_elements(By.XPATH, "//div[@class='basket_content_item']"))
        close_button.click()
        return count

    def click(self, elem: WebElement, header: WebElement = None):
        if header is None:
            header = self.driver.find_element(By.XPATH, "//header")

        ActionChains(self.driver).scroll_by_amount(0, int(elem.rect["y"] - header.rect["height"])).perform()
        elem.click()

    def check_item_exists_in_shopcart(self, elem: WebElement):
        self.driver.find_element(*self.basket_icon).click()
        close_button: WebElement = WebDriverWait(self.driver, 15).until(ec.element_to_be_clickable((By.XPATH, "//button[@id=\"basketCloseButton\"]")))
        shopcart_items: List[WebElement] = self.driver.find_elements(By.XPATH, "//div[@class=\"basket_content_item\"]")

        for item in shopcart_items:
            if self.item_to_value(elem) == self.basket_item_to_value(item):
                close_button.click()
                return True

        close_button.click()
        return False

    @property
    def basket_items_price(self) -> int:
        self.driver.find_element(*self.basket_icon).click()
        close_button: WebElement = WebDriverWait(self.driver, 15).until(ec.element_to_be_clickable((By.XPATH, "//button[@id=\"basketCloseButton\"]")))

        shopcart_items: List[WebElement] = self.driver.find_elements(By.XPATH, "//div[@class=\"basket_content_item\"]//p/span[1]")
        total_price: int = 0
        for item in shopcart_items:
            total_price += int("".join(item.text.split(",")))

        close_button.click()
        return total_price

    def item_to_value(self, elem: WebElement) -> Dict[str, str]:
        return {
            "name": elem.find_element(By.XPATH, ".//p/a").text,
            "price": elem.find_element(By.XPATH, ".//div/p/span/span").text
        }

    def basket_item_to_value(self, elem: WebElement):
        item_properties: Dict[str, str] = {
            "name": elem.find_element(By.XPATH, ".//h4").text,
            "price": elem.find_element(By.XPATH, ".//p/span[1]").text
        }
        return item_properties

